demo/mbtaupsert/src/main.rs

```diff
@@ -0,0 +1,286 @@
+    let mut reader = BufReader::new(f);
+    let mut timestamp = 0;
+    // tracks bytes read in the previous iteration of the loop.
+    // this is only used when there is an error parsing the line
+    // sometimes, the log parser is overly eager and reads a line before
+    // it has finished writing. so we want to retry reading the whole line
+    // until number of bytes read from the line is no longer increasing
+    let mut old_len = 0;
+
+    let exit_at_end = opts.opt_present("exit-at-end");
+
+    loop {
+        let mut line = String::new();
+        let resp = reader.read_line(&mut line);
+        match resp {
+            Ok(len) => {
+                if len > 0 {
+                    match parse_line(&line) {
+                        Ok(key_values) => {
+                            for (key, value) in key_values {
+                                let mut partition_key = (calculate_hash(&key) as i32) % partitions;
+                                if partition_key < 0 {
+                                    partition_key += partitions;
+                                }
+                                state.ingest(&topic_name, partition_key, Some(key), value)?;
+                                if byo {
+                                    state.ingest_consistency(
+                                        &consistency_name,
+                                        &topic_name,
+                                        timestamp,
+                                        timestamp + 1,
+                                    )?;
+                                }
+                                timestamp += 1;
+                            }
+                            old_len = 0;
+                            pos += len as u64;
+                            reader.seek(SeekFrom::Start(pos)).unwrap();
+                        }
+                        Err(msg) => {
+                            if len <= old_len {
+                                // emit an error and advance the reader.
+                                println!("{}", msg);
+                                pos += len as u64;
+                                reader.seek(SeekFrom::Start(pos)).unwrap();
+                            } else {
+                                //retry after waiting for the line to update.
+                                old_len = len;
+                                thread::sleep(time::Duration::from_millis(250));
+                            }
+                        }
+                    }
+                    line.clear();
+                } else {
+                    if exit_at_end {
+                        return Ok(());
+                    }
+                    thread::sleep(heartbeat);
+                }
+            }
+            Err(err) => {
+                println!("{}", err);
+            }
+        }
+    }
+}
+
+fn main() {
+    match run_stream() {
+        Ok(()) => {}
+        Err(err) => println!("{}", err),
```

Comment:
This violates standard Unix conventions for error handling. The error should be printed to stderr, not stdout, and the return code of the program should be nonzero.

If you don't exit with a failing status code, then CI won't be able to check whether this demo succeeded or not.